### PR TITLE
DKG nuke command

### PR DESCRIPTION
**Pull request by CluEleSsUK**
DKG nuke command to easily delete a single beacon's state, should you need to

### CODE REVIEW

**Comment by CluEleSsUK** [comment]
I could add some mickey mouse tests to bump up the coverage but it seems a waste of time

File: internal/dkg/store.go [hunk 1]
```diff
@@ -17,8 +17,8 @@ import (
 	"github.com/drand/drand/v2/common/log"
 )
 
-type boltStore struct {
-	sync.RWMutex
+type BoltStore struct {
+	lock          sync.RWMutex
```
Comment (AnomalRoil) [inline internal/dkg/store.go:21]: why the need for explicit field for it?

File: internal/dkg/store.go [hunk 2]
```diff
@@ -225,3 +225,14 @@ func (s *boltStore) MigrateFromGroupfile(beaconID string, groupFile *key.Group,
 
 	return s.SaveFinished(beaconID, &state)
 }
+
+func (s *BoltStore) NukeState(beaconID string) error {
```
Comment (AnomalRoil) [inline internal/dkg/store.go:229]: Why not lock it during the process here?

**Review by AnomalRoil — CHANGES_REQUESTED** [review]

File: internal/dkg/state_machine.go [hunk 3]
```diff
@@ -626,7 +626,7 @@ var ErrLeaderCantJoinAfterFirstEpoch = errors.New("you cannot lead a DKG and joi
 var ErrLeaderNotRemaining = errors.New("you cannot lead a DKG and leave at the same time")
 var ErrLeaderNotJoining = errors.New("the leader must join in the first epoch")
 var ErrOnlyJoinersAllowedForFirstEpoch = errors.New("participants can only be joiners for the first epoch")
-var ErrNoNodesRemaining = errors.New("cannot propose a network common.Without nodes remaining")
+var ErrNoNodesRemaining = errors.New("cannot propose a network without nodes remaining")
```
[the remark this answers is not in the document]
Comment (CluEleSsUK) [reply internal/dkg/state_machine.go:629]: I think it was a find+replace error from a previous merge

File: internal/dkg/store.go (hunk 1 above)
Comment (CluEleSsUK) [reply internal/dkg/store.go:21]: linter complained about embedding Mutex in public structs

File: internal/dkg/state_machine.go [hunk 4]
```diff
@@ -689,8 +691,35 @@ func hasTimedOut(details *DBState) bool {
...
+	}
+
+	if err := validateReshareTerms(terms); err != nil {
+		return err
+	}
+
+	// nodes joining after the first epoch accept some things at face value
+	// nodes already in the network shouldn't accept e.g. a change of genesis time
+	if currentState.State != Fresh {
+		err = validateReshareContinuity(currentState, terms)
```
[the remark this answers is not in the document]
Comment (CluEleSsUK) [reply internal/dkg/state_machine.go:713]: damn, I struggled for ages and thought this name was a breakthrough haha - what would you prefer?

File: internal/dkg/store.go (hunk 2 above)
Comment (CluEleSsUK) [reply internal/dkg/store.go:229]: taking a second look at the code, perhaps it can be entirely removed - it maintains a fake transaction between two writes on `SaveFinished` but in fact that is already supported out-of-the-box with bolt

File: internal/dkg/state_machine.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (AnomalRoil) [reply internal/dkg/state_machine.go:713]: What does it do? It validates reshare as a new joiner? `validateReshareNewJoiner` maybe?

File: internal/dkg/store.go (hunk 1 above)
Comment (AnomalRoil) [reply internal/dkg/store.go:21]: Ah right, does this one really need to be public? Although it's in our internal folder, so it's fine. We can had a rule to our linter to accept these in `internal` I think.

Comment (CluEleSsUK) [reply internal/dkg/store.go:21]: `unexported-return: exported func NewDKGStore returns unexported type *dkg.boltStore, which can be annoying to us` I think because I use it in the dkg cli

File: internal/dkg/state_machine.go [hunk 5]
```diff
@@ -770,23 +775,44 @@ func validateEpoch(currentState *DBState, terms *drand.ProposalTerms) error {
...
 		return ErrLeaderNotJoining
 	}
 	return nil
 }
 
-func validateReshare(currentState *DBState, terms *drand.ProposalTerms) error {
+func validateReshareTerms(terms *drand.ProposalTerms) error {
+	if len(terms.Remaining) == 0 {
+		return ErrNoNodesRemaining
+	}
```
[the remark this answers is not in the document]
Comment (CluEleSsUK) [reply internal/dkg/state_machine.go:800]: there's a test if the threshold is too high, but not the other way round. I imagine the DKG will just fail, but it's a good idea to add an early check - will do

File: internal/core/drand_test.go [hunk 6]
```diff
@@ -309,7 +309,8 @@ func TestRunDKGReshareAbsentNodeDuringExecution(t *testing.T) {
 			t.Logf("Node %d stopped \n", nodeIndexToStop)
 		},
 	}
-	newGroup, err := dt.RunReshareWithHooks(t, dt.clock.Now().Add(3*beaconPeriod), currentNodes, newNodes, hooks)
+	dt.thr = 3
```
Comment (AnomalRoil) [inline internal/core/drand_test.go:312]: Why this change? Isn't that set in `NewDrandTestScenario`? 
Also, why not use one of the threshold variable declared above?

File: internal/dkg/state_machine.go [hunk 7]
```diff
@@ -740,10 +743,19 @@ func validateForAllDKGs(currentState *DBState, terms *drand.ProposalTerms) error
 		return ErrTimeoutReached
 	}
 
-	if int(terms.Threshold) > len(terms.Joining)+len(terms.Remaining) {
+	nodeCount := len(terms.Joining) + len(terms.Remaining)
+	if int(terms.Threshold) > nodeCount {
 		return ErrThresholdHigherThanNodeCount
 	}
 
+	if nodeCount < int(currentState.Threshold) {
```
Comment (AnomalRoil) [inline internal/dkg/state_machine.go:751]: Shouldn't that be rather
```suggestion
	if len(terms.Remaining) < int(currentState.Threshold) {
```

File: internal/core/drand_test.go (hunk 6 above)
Comment (CluEleSsUK) [reply internal/core/drand_test.go:312]: you pass it into the `NewDrandTestScenario`, but it doesn't get updated for reshares

Comment (AnomalRoil) [reply internal/core/drand_test.go:312]: Would the proper fix here be to adapt the threshold in the `RunReshareWithHooks` function?

File: internal/dkg/state_machine.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (CluEleSsUK) [reply internal/dkg/state_machine.go:716]: well it's actually the opposite, new joiners don't validate this stuff, just existing members

File: internal/dkg/state_machine.go (hunk 7 above)
Comment (CluEleSsUK) [reply internal/dkg/state_machine.go:751]: I changed this but github doesn't show outdated, not sure why

File: internal/core/drand_test.go (hunk 6 above)
Comment (CluEleSsUK) [reply internal/core/drand_test.go]: done

File: internal/dkg/state_machine.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (CluEleSsUK) [reply internal/dkg/state_machine.go]: have changed the name to reflect it

File: internal/dkg/state_machine.go [hunk 8]
```diff
@@ -679,7 +682,9 @@ func isValidStateChange(current, next Status) bool {
 	case TimedOut:
 		return next == Proposing || next == Proposed || next == Aborted
 	case Failed:
-		return next == Proposing || next == Proposed || next == Joined || next == Left || next == Aborted
+		// a node can be `Failed` but still be included in the group file under some (magical) circumstances.
+		// In such a case, it should be added as a remainer on the next DKG rather than a joiner.
```
Comment (AnomalRoil) [inline internal/dkg/state_machine.go:686]: I kinda disagree: if a node failed to join, they don't have a valid share and so cannot act as a proper "remainer" in the DKG. 
Are we sure such a node is going to use the "new joiner" codepath despite them being "remainer" here?

File: internal/dkg/state_machine.go (hunk 7 above)
Comment (AnomalRoil) [reply internal/dkg/state_machine.go:751]: I think my comment still holds:

if there aren't enough remaining nodes (+ leader) in the proposal, we cannot expect to run a successful DKG since we need at least threshold nodes with a secret share to do a resharing (and since currently the leavers aren't participating in the resharing IIRC the code)

File: internal/dkg/state_machine.go [hunk 9]
```diff
@@ -689,69 +694,81 @@ func hasTimedOut(details *DBState) bool {
...
-		return validateReshare(currentState, terms)
+func validateJoinerSignatures(terms *drand.ProposalTerms, targetSch *crypto.Scheme) error {
+	for _, participant := range terms.Joining {
+		id, err := key.IdentityFromProto(participant, targetSch)
+		if err != nil {
+			return fmt.Errorf("%w, participant error: %s, expected: %s", key.ErrInvalidKeyScheme, err.Error(), targetSch.Name)
+		}
+		if err := id.ValidSignature(); err != nil {
+			return key.ErrInvalidKeyScheme
+		}
```
Comment (AnomalRoil) [inline internal/dkg/state_machine.go:770]: Why don't we validate the signatures of node remaining? That seems like an attack vector for a malicious leader: sneak invalid signatures/public key pairs in there and bam you can take down the network after a resharing?

I guess a secondary problem here is that we "trust" signatures and keys coming from the same proposal, so as a malicious leader I could create fake keys and include valid signatures anyway. 

Not sure what kind of problems that would cause down the line, FWIW, I think we could even get rid of the long term public keys:  after we've joined we can use our secret share to sign messages and that's impossible to fake.

File: internal/dkg/state_machine.go (hunk 7 above)
Comment (AnomalRoil) [reply internal/dkg/state_machine.go:751]: I think you only changed it below in `validateReshareTerms`

**Review by AnomalRoil — APPROVED** [review]

File: internal/dkg/state_machine.go (hunk 7 above)
Comment (CluEleSsUK) [reply internal/dkg/state_machine.go:751]: ah oops, my find/replace was on regex mode fuuu

File: internal/dkg/state_machine.go (hunk 8 above)
Comment (CluEleSsUK) [reply internal/dkg/state_machine.go:686]: an example could be if they have a network timeout; their node will be in a failed state, though the other nodes will seem in the group file. 
Given they're in the group file, if another user proposes a new DKG, all the other nodes will expect them to be a `Remainer`. Unfortunately "alive for the DKG" is completely separate from "send a valid partial at any point after the DKG".
The failed node won't need to use the new joiner path as they'll still have the node list from previous steps

File: internal/dkg/state_machine.go (hunk 9 above)
Comment (CluEleSsUK) [reply internal/dkg/state_machine.go:770]: checking if a node ought to be remaining (ie. it was in a previous epoch) compares the key and signatures; to that end, a leader modifying a signature would appear as a `Remainer` not present in the prior epoch and fail.
The only route for this being an attack vector I can see would be on the DKG from v1.5.9 -> v2, as technically the sigs aren't validated because nobody was ever a `Joiner`. That said, we have no new parties joining
